### PR TITLE
Espresso migration

### PR DESCRIPTION
**Pull request by zacshowa**
Closes # [162](https://github.com/EspressoSystems/nitro-espresso-integration/issues/162)
### TODO
- [x] Clean up script, add clear comments on what steps are doing what. 
- [x] Explain which steps are really part of the Upgrade process and which are only to make the test work.
- [x] Add a README.md,
- [x] Verify staker is working, depends on https://github.com/EspressoSystems/nitro-espresso-integration/pull/210
- [x] Do not use a docker image built from a Pull Request (i. e. merge changes we need into `integration` branch on our nitro fork)

### This PR:

Is a draft for a PR that will include a full E2E migration test using the nitro test node.
Currently this PR:
 - starts a default nitro sequencer via the included tooling, including setting up a rollup.
 - starts an instance of the espresso-dev-node
 - shuts down the default nitro sequencer node
 - attempts to start the Espresso integrated nitro sequencer node in a new docker image using the same storage volumes, with a different config volume. (I'm thinking out loud at the time of writing this description. Maybe it makes more sense to do something different with the same volume instead of making a new volume.  I was worried if the volume was still in use, I wouldn't be able to write to it.)
 

### This PR does not:

Currently install the new contracts and attempt to perform the upgrade.

### Key places to review:

the espresso integrated sequencer is currently not starting due to a lack of some required items in its config. I still need to track down and copy these over to the correct volume.

Run the script migration-test.bash

### CODE REVIEW

File: docker-compose.yaml [hunk 1]
```diff
@@ -165,6 +165,23 @@ services:
     depends_on:
       - geth
 
+  espresso-sequencer:
+    pid: host # allow debugging
+    image: espresso-integration-testnode
```
Comment (ImJeremyHe) [inline docker-compose.yaml:170]: what's the difference between this and `nitro-node-dev-testnode`

File: scripts/config.ts [hunk 2]
```diff
@@ -29,6 +29,8 @@ SLOTS_PER_EPOCH: 6
 DEPOSIT_CONTRACT_ADDRESS: 0x4242424242424242424242424242424242424242
     `
     fs.writeFileSync(path.join(consts.configpath, "prysm.yaml"), prysm)
+    fs.writeFileSync(path.join(consts.espressoconfigpath, "prysm.yaml"), prysm)
```
Comment (sveitser) [inline scripts/config.ts]: I think rather than writing 2 files from this script we should call the script twice with different `configpath` (and arguments).

File: docker-compose.yaml [hunk 3]
```diff
@@ -165,6 +165,23 @@ services:
     depends_on:
       - geth
 
+  espresso-sequencer:
```
Comment (sveitser) [inline docker-compose.yaml]: Maybe `sequencer-on-espresso`. Otherwise it's easily confused with the Espresso sequencer.

File: docker-compose.yaml (hunk 1 above)
Comment (sveitser) [reply docker-compose.yaml:170]: I think this should be `ghcr.io/espressosystems/nitro-espresso-integration/nitro-node-dev:integration` eventually but it's currently necessary to use a local docker image for testing because we haven't merged @zacshowa's E2E changes in nitro and they will be needed for this upgrade test. So I guess it's a locally tagged image.

File: scripts/config.ts (hunk 2 above)
Comment (sveitser) [reply scripts/config.ts]: A hacky way to do is I think is to create a second `scripts-espresso` service like the `scripts` service and have that one mount `/config-espresso:/config`. Then we don't need to make more edits to this script.

Something like

```yaml
  scripts-espresso:
    build: scripts/
    volumes:
      - "l1keystore:/home/user/l1keystore"
      - "config-espresso:/config"
      - "tokenbridge-data:/tokenbridge-data"
```

cc @zacshowa @ImJeremyHe

File: espresso-tests/.env [hunk 4]
```diff
@@ -0,0 +1,21 @@
+# env vars for chain name and rpc_url
+CHAIN_NAME="1337"
```
Comment (sveitser) [inline espresso-tests/.env]: `CHAIND_ID`, also we should prefix everything from L1 with `L1`, or `PARENT_CHAIN`.

File: espresso-tests/migration-test.bash [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+
+echo "Deploying ArbOS action"
+
+#forge script to deploy the Espresso ArbOS upgrade action.
+cd $ORBIT_ACTIONS_DIR
+forge script --chain $L2_CHAIN_NAME contracts/child-chain/arbos-upgrade/DeployArbOSUpgradeAction.s.sol:DeployArbOSUpgradeAction  --rpc-url $L2_RPC_URL --broadcast -vvvv
+ARBOS_UPGRADE_ACTION=$(cat broadcast/DeployArbOSUpgradeAction.s.sol/412346/run-latest.json | jq -r '.transactions[0].contractAddress')
+echo "Deployed ArbOSUpgradeAction at $ARBOS_UPGRADE_ACTION"
+
+# TODO: figure out why this upgrade does not do anything
```
Comment (sveitser) [inline espresso-tests/migration-test.bash]: This comment can be removed now.

File: espresso-tests/migration-test.bash [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+
+# Export l2 owner private key and address
+PRIVATE_KEY="$(docker compose run scripts print-private-key --account l2owner | tail -n 1 | tr -d '\r\n')"
+OWNER_ADDRESS="$(docker compose run scripts print-address --account l2owner | tail -n 1 | tr -d '\r\n')"
+
+
+cd "$ORBIT_ACTIONS_DIR"
+
+git submodule update --init
+yarn
```
Comment (sveitser) [inline espresso-tests/migration-test.bash]: I think it's a bit weird to run these installation commands in the middle of the script. Let's move them to the start or out of this script.

File: espresso-tests/migration-test.bash [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+# Wait for RPC_URL to be available
+while ! curl -s $L2_RPC_URL > /dev/null; do
+  echo "Waiting for $L2_RPC_URL to be available..."
+  sleep 5
+done
+
+echo "Adding child chain upgrade executor as an L2 chain owner"
+cast send 0x0000000000000000000000000000000000000070 'addChainOwner(address)' $CHILD_CHAIN_UPGRADE_EXECUTOR_ADDRESS --rpc-url $L2_RPC_URL --private-key $PRIVATE_KEY
+
+echo "Deploying ArbOS action"
```
Comment (sveitser) [inline espresso-tests/migration-test.bash]:
```suggestion
echo "Deploying ArbOS Upgrade action"
```

File: espresso-tests/migration-test.bash [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+#forge script to deploy the Espresso ArbOS upgrade action.
+cd $ORBIT_ACTIONS_DIR
+forge script --chain $L2_CHAIN_NAME contracts/child-chain/arbos-upgrade/DeployArbOSUpgradeAction.s.sol:DeployArbOSUpgradeAction  --rpc-url $L2_RPC_URL --broadcast -vvvv
+ARBOS_UPGRADE_ACTION=$(cat broadcast/DeployArbOSUpgradeAction.s.sol/412346/run-latest.json | jq -r '.transactions[0].contractAddress')
+echo "Deployed ArbOSUpgradeAction at $ARBOS_UPGRADE_ACTION"
+
+# TODO: figure out why this upgrade does not do anything
+cast send $CHILD_CHAIN_UPGRADE_EXECUTOR_ADDRESS "execute(address, bytes)" $ARBOS_UPGRADE_ACTION $(cast calldata "perform()") --rpc-url $L2_RPC_URL --private-key $PRIVATE_KEY
+
+ARBOS_VERSION_BEFORE_UPGRADE=$(cast call "0x0000000000000000000000000000000000000064" "arbOSVersion()(uint64)" --rpc-url $L2_RPC_URL)
```
Comment (sveitser) [inline espresso-tests/migration-test.bash]: Should be moved before the cast call above.

File: espresso-tests/migration-test.bash [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+
+# The arbsys precompile is returning 55 for the arbos version which indicates that the value internally is 0
+# We are upgrading the ArbOS version to 35 so the expect the return value to be 55 + 35 = 90
+if [ $ARBOS_VERSION_AFTER_UPGRADE != "90" ]; then
+  fail "ArbOS version not updated: Expected 90, Actual $ARBOS_VERSION_AFTER_UPGRADE"
+fi
+
+#test for new OSP address
+CHALLENGE_MANAGER_OSP_ADDRESS=$(cast call $CHALLENGE_MANAGER_ADDRESS "osp()(address)" --rpc-url $RPC_URL)
+if [ $NEW_OSP_ENTRY != $CHALLENGE_MANAGER_OSP_ADDRESS ]; then
```
Comment (sveitser) [inline espresso-tests/migration-test.bash:149]: This currently fails because the address from the broadcast artifact is not checksummed (it's all lowercase) whereas the address we get from cast is checksummed.

Comment (sveitser) [reply espresso-tests/migration-test.bash:149]: Can use `cast to-checksum` E. g. 

```
cat orbit-actions/broadcast/DeployEspressoOsp.s.sol/1337/run-latest.json | jq -r '.transactions[4].contractAddress' | cast to-checksum
```

Comment (zacshowa) [reply espresso-tests/migration-test.bash:149]: Do we need to do this for the other addresses that we grab from the deployment files?

File: espresso-tests/.env (hunk 4 above)
Comment (zacshowa) [reply espresso-tests/.env]: I think renaming everything to parent and child chain makes sense as L1 and L2 don't necessarily make sense in the context of orbit chains.

File: espresso-tests/migration-test.bash (hunk 9 above)
Comment (sveitser) [reply espresso-tests/migration-test.bash:149]: We only need to do it if we compare addresses as strings. Alternatively we can convert everything to lowercase.

File: espresso-tests/.env [hunk 10]
```diff
@@ -1,21 +1,23 @@
-# env vars for chain name and rpc_url
-CHAIN_NAME="1337"
-L2_CHAIN_NAME="412346"
-RPC_URL="http://localhost:8545"
-L2_RPC_URL="http://localhost:8547"
-# env vars for new OSP deployment
+# Environment variables for chain name and rpc_url
+# These are essential for the upgrade
+PARENT_CHAIN_CHAIN_ID="1337"
+CHILD_CHAIN_CHAIN_NAME="412346"
```
Comment (sveitser) [inline espresso-tests/.env:4]: I there a reason to call this "name"? It's also the chain ID.

File: espresso-tests/README.md [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+This document contains details surrounding the test migrating an exsisting Arbitrum Orbit chain to use the nitro-espresso-integration, and be compatible with the Espresso network.
+
+## Running the test
+
+Run the test by navigating to nitro-testnode/espresso-tests and run ./migration-test.bash
+
+
+## Steps related to upgrade in production
+
+Many of the steps in the test are nearly identical to the steps needed to upgrade an orbit network to be compatible with the espresso network.
```
Comment (sveitser) [inline espresso-tests/README.md:12]: I think the instructions on how to carry out the upgrade should go into a new README in `orbit-actions`, not this repo. They should also contain instructions to deploy and execute all the actions required for the Espresso upgrade.

Let's link to this test from the new README in orbit-actions for the curious.

File: espresso-tests/.env (hunk 10 above)
Comment (zacshowa) [reply espresso-tests/.env:4]: No, we can rename this. In my head chain name and chain id are interchangeable which is not the case for Arbitrum or Ethereum.

File: espresso-tests/README.md (hunk 11 above)
Comment (zacshowa) [reply espresso-tests/README.md:12]: I agree and I'm actively working on a README in that repo.

File: test-node.bash [hunk 12]
```diff
@@ -433,11 +444,17 @@ if $force_init; then
...
     docker compose run --entrypoint sh rollupcreator -c "jq [.[]] /config/deployed_chain_info.json > /config/l2_chain_info.json"
+    docker compose run --entrypoint sh rollupcreator -c "jq [.[]] /config/deployed_chain_info.json > /espresso-config/l2_chain_info.json"
     docker compose run --entrypoint sh rollupcreator -c "cat /config/l2_chain_info.json"
 
     if $simple; then
-        echo == Writing configs
-        docker compose run scripts write-config --simple
+        echo == Writing configs for simple
+        docker compose run scripts write-config --simple --simpleWithValidator $simple_with_validator --espresso $l2_espresso --lightClientAddress $lightClientAddr
+        if $l2_espresso; then
```
Comment (ImJeremyHe) [inline test-node.bash]: Can you check if this is necessary? The only difference is running a different docker container but `scripts` and `scripts-espresso` actually are the same? And I think we need to use `scripts-espresso` only in the migration test, not the test-node.bash?

File: scripts/index.ts [hunk 13]
```diff
@@ -2,7 +2,7 @@ import { hideBin } from "yargs/helpers";
 import Yargs from "yargs/yargs";
 import { stressOptions } from "./stress";
 import { redisReadCommand, redisInitCommand } from "./redis";
-import { writeConfigCommand, writeGethGenesisCommand, writePrysmCommand, writeL2ChainConfigCommand, writeL3ChainConfigCommand } from "./config";
+import { writeConfigCommand, writeGethGenesisCommand, writePrysmCommand, writeL2ChainConfigCommand, writeL3ChainConfigCommand} from "./config";
```
Comment (ImJeremyHe) [inline scripts/index.ts]: Unnecessary changes

File: scripts/config.ts [hunk 14]
```diff
@@ -489,4 +531,4 @@ export const writeL3ChainConfigCommand = {
     handler: (argv: any) => {
         writeL3ChainConfig(argv)
     }
-}
+}
```
Comment (ImJeremyHe) [inline scripts/config.ts:543]: Unnecessary changes

File: scripts/Dockerfile [hunk 15]
```diff
@@ -3,5 +3,8 @@ WORKDIR /workspace
 COPY ./package.json ./yarn.lock ./
 RUN yarn
 COPY ./*.ts ./tsconfig.json ./
+RUN mkdir /config
+# populate this file with valid json that will later be overwritten. If we don't do this the docker image won't build because of typescript looking for this file before the volume that contains it is mounted (I assume)
+RUN cp ./tsconfig.json /config/l2_chain_info.json
```
Comment (ImJeremyHe) [inline scripts/Dockerfile]: Is this still an issue? A bit strange I think

File: scripts/config.ts (hunk 14 above)
Comment (zacshowa) [reply scripts/config.ts:543]: I agree that this is unnecessary, but I don't know what I would do to fix this part of the commit log. I had added and removed a function after this brace. However, I didn't actually modify it at all, so I'm not exactly sure why git is registering it as a changed line.

File: test-node.bash (hunk 12 above)
Comment (zacshowa) [reply test-node.bash]: I do think that this should be moved to the migration test bash script. I just remember this being a cleaner way to write a config to a different config volume than to write to two different locations in the javascript code. I believe it's probably better to write the configs to two different volumes as well. IMO overwriting the original config volume would be less clear to any future readers of the code.

File: scripts/config.ts (hunk 14 above)
Comment (ImJeremyHe) [reply scripts/config.ts:543]: Are you using some formatter extension to remove the trailing new line?

**Review by ImJeremyHe — APPROVED** [review]
